feat(invariant): assert all invariants by default (`assert_all`)

grandizzy opened the pull request:
<!-- ** Please select "Allow edits from maintainers" in the PR Options ** -->

closes #9727

Adds `invariant.assert_all` (default `true`, matches Echidna/Medusa). When enabled the campaign keeps running after the first broken invariant and shrinks every broken one in a single run.

What this PR adds:
- Per-invariant rendering: each broken invariant gets a `[FAIL: reason]` line + `[Sequence]` block. The invariant function name is appended to the FAIL line whenever it's needed to disambiguate — i.e. when there are multiple failures, or the single failure isn't the campaign anchor. The anchor's name is omitted in single-failure runs because the trailing `<name>() (runs: …)` summary already identifies it.
- `Suite assert_all: <broken>/<total> invariants broken` roll-up above the per-invariant blocks
- `[i/N] Shrink: <invariant>` counter on the progress bar when more than one invariant needs shrinking
- `N invariant failures persisted to <dir> — rerun to shrink` footer
- On Ctrl+C: every known broken invariant is reported and the un-shrunk sequence is persisted, so re-running targets that invariant shrinks from cache
- Warnings: `assert_all + optimization invariant` (boolean secondaries silently filtered) and `N invariant(s) skipped due to persisted failures` on re-runs
- Set `invariant.assert_all = false` to restore the legacy "stop on first broken invariant" behavior

Internal cleanup behind the new behavior: a single `invariant_fns` list per `InvariantContract` (with an `anchor_idx`), a single `invariant_failures: Vec<InvariantFailure>` on `TestResult`, and direct return of the broken invariant from the assertion path (removing the prior `first_broken_event` re-scan). The campaign **anchor** is the `--mt`-selected invariant (or the only one), shrunk and persisted at the campaign root path; secondaries share the same shape and are shrunk + persisted under their own filenames.

Parameterized `invariant_*` functions (e.g. `invariant_bar(uint256 z)`) are now rejected upstream with a per-function `TestResult::fail` entry, instead of silently leaking into `assert_all` secondaries and breaking campaigns mid-run with confusing decode errors.

@forkforkdog @rappie @GalloDaSballo appreciate some feedback and if that's what you were looking for, thank you
CC @0xalpharush

## PR Checklist

- [x] Added Tests
- [x] Added Documentation
- [x] Breaking changes (default config behavior change: campaigns now report all broken invariants by default)

grandizzy committed:
feat(invariant): assert all invariants

DaniPopes commented on crates/config/src/invariant.rs:
```diff
@@ -37,6 +37,8 @@ pub struct InvariantConfig {
     pub timeout: Option<u32>,
     /// Display counterexample as solidity calls.
     pub show_solidity: bool,
+    /// Continue invariant run until all invariants declared in current test suite breaks.
```
```suggestion
    /// Continue invariant run until all invariants declared in the test suite break.
```

0xalpharush commented:
> invariants that already have persisted failure are not picked up / asserted

Not to say this can't be done, but this isn't how Medusa/Echidna do it. The "pro" that comes to mind is you may find a smaller reproducer or a reproducer that reveals the patch was incomplete i.e. only fixed one repro but not the other.

It is possible to do deduplication with the edge map to avoid saving too many if you went this route.

grandizzy commented:
> > invariants that already have persisted failure are not picked up / asserted
> 
> Not to say this can't be done, but this isn't how Medusa/Echidna do it. The "pro" that comes to mind is you may find a smaller reproducer or a reproducer that reveals the patch was incomplete i.e. only fixed one repro but not the other.
> 
> It is possible to do deduplication with the edge map to avoid saving too many if you went this route.

I see, makes sense, can accommodate such

rappie commented:
This looks great. Some quick ideas for further improvements:
- Keep running even after all invariants a broken because there might still be unbroken assertions in handlers
- Keep track of broken handlers. Once a handler is broken once, ignore all its assertions BUT keep calling it for building interesting sequences

Kushmanmb commented:
Let's not play micky mouse games. Could man to man ask. That honestly more logical

grandizzy committed:
Tests and Nits

Amp-Thread-ID: https://ampcode.com/threads/T-019dbf48-3fb0-7762-a01f-b5e966339e73
Co-authored-by: Amp <amp@ampcode.com>

grandizzy committed:
fix: check all invariants in afterInvariant gate and preflight

Amp-Thread-ID: https://ampcode.com/threads/T-019dbf48-3fb0-7762-a01f-b5e966339e73
Co-authored-by: Amp <amp@ampcode.com>

grandizzy committed:
fix: use per-invariant fail_on_revert when recording handler revert failures

Co-authored-by: Amp <amp@ampcode.com>
Amp-Thread-ID: https://ampcode.com/threads/T-019dbf48-3fb0-7762-a01f-b5e966339e73

grandizzy committed:
fix: commit state between txs in generate_counterexample

Amp-Thread-ID: https://ampcode.com/threads/T-019dbf48-3fb0-7762-a01f-b5e966339e73
Co-authored-by: Amp <amp@ampcode.com>

grandizzy committed:
fix: preflight check all invariants, not just primary

Amp-Thread-ID: https://ampcode.com/threads/T-019dbf48-3fb0-7762-a01f-b5e966339e73
Co-authored-by: Amp <amp@ampcode.com>

grandizzy committed:
fix: exclude secondary invariants from optimization mode runs

Amp-Thread-ID: https://ampcode.com/threads/T-019dbf48-3fb0-7762-a01f-b5e966339e73
Co-authored-by: Amp <amp@ampcode.com>

grandizzy committed:
refactor: rename invariant_fn to primary_invariant_fn, deterministic preflight error, debug_assert on empty invariants

Amp-Thread-ID: https://ampcode.com/threads/T-019dbf48-3fb0-7762-a01f-b5e966339e73
Co-authored-by: Amp <amp@ampcode.com>

grandizzy committed:
feat: show broken invariant count in progress bar during continuous runs

Amp-Thread-ID: https://ampcode.com/threads/T-019dbf48-3fb0-7762-a01f-b5e966339e73
Co-authored-by: Amp <amp@ampcode.com>

grandizzy committed:
feat(invariant): rename continuous_run to assert_all and default to true

Renames the InvariantConfig field to better describe its semantics
("assert every invariant in the suite, don't stop on first failure")
and flips the default to true so multi-invariant suites report all
broken invariants by default, matching Echidna/Medusa behavior.

Co-authored-by: Amp <amp@ampcode.com>
Amp-Thread-ID: https://ampcode.com/threads/T-019dcd68-66ac-76ed-ac5c-7ea722a9c9ae

grandizzy committed:
feat(invariant): parameterize shrinker by target invariant + persisted failures footer

- Generalize shrink_sequence, shrink_sequence_value, replay_run, replay_error
  to accept target_invariant: &Function (currently always primary; unblocks
  per-secondary shrinking).
- Move reset_shrink_progress out of shrink fns; called once per invariant from
  replay_error. Progress label now 'Shrink: <invariant_name>'.
- Add TestResult.invariant_failure_dir; Display appends
  'N invariant failures persisted to <dir> — rerun to shrink' when secondary
  failures were written.

Amp-Thread-ID: https://ampcode.com/threads/T-019dcd68-66ac-76ed-ac5c-7ea722a9c9ae
Co-authored-by: Amp <amp@ampcode.com>

grandizzy committed:
feat(invariant): structured InvariantOtherFailure for assert_all secondaries

Promotes TestResult.other_failures from Vec<String> to Vec<InvariantOtherFailure>
carrying name, reason, optional counterexample, and persisted path. Display
renders each secondary symmetrically with [FAIL: reason] + [Sequence] block
when a counterexample is available, falling back to the terse 'name: reason'
one-liner otherwise.

Amp-Thread-ID: https://ampcode.com/threads/T-019dcd68-66ac-76ed-ac5c-7ea722a9c9ae
Co-authored-by: Amp <amp@ampcode.com>

grandizzy committed:
feat(invariant): serial secondary shrinking + Ctrl-C persists un-shrunk secondaries

PR-3 of the assert_all rollout. After the campaign finishes, every broken
secondary invariant is shrunk in turn via replay_error so users get a
ready-to-debug counterexample for each failure in a single run (matching
how the primary is rendered: [FAIL: reason] <name> + [Sequence] block).

On Ctrl-C, instead of dropping known secondaries (previous behavior was a
'break' before pushing them), the loop keeps recording every failure the
campaign discovered. The shrink + replay step is skipped to honor the
interrupt, but the un-shrunk sequence is persisted via
BaseCounterExample::from_invariant_call (no execution required), so a
re-run targeting that secondary picks up the saved counterexample and
shrinks from there — same UX as re-running an interrupted primary.

Output of an interrupted run now includes a terse
'<invariant>: <reason>' line for each secondary the campaign saw,
preserving visibility of all broken invariants while keeping the
interrupt fast.

Adds e2e coverage:
- assert_all: extended to verify secondary failures render symmetrically
  with shrunk sequences and that re-running skips persisted secondaries.
- assert_all_only_primary: new test confirming no secondary [FAIL] blocks
  or persisted-failures footer appear when only the primary breaks.

Co-authored-by: Amp <amp@ampcode.com>
Amp-Thread-ID: https://ampcode.com/threads/T-019dcdd3-53f5-76b6-ac36-d59f06b58280

grandizzy committed:
feat(invariant): assert_all polish — [i/N] shrink counter, suite roll-up, opt-mode warning

Three small UX wins for assert_all campaigns. No behavior change, no
new dependencies.

1. Shrink progress bar gets an [i/N] queue counter when more than one
   invariant needs shrinking, so users see how many shrinkers are
   queued behind the current one (e.g. '[2/3] Shrink: invariant_X').
   reset_shrink_progress and replay_error gain a position parameter;
   single-invariant call sites pass None.

2. Suite-level roll-up footer: when assert_all exercised >1 invariant
   and the test failed, render 'Suite assert_all: <broken>/<total>
   invariants broken' above the per-invariant blocks. Gives CI logs
   and Slack pastes a glanceable health line. New Option<usize>
   field on TestResult, populated only when meaningful.

3. Startup warning when assert_all + optimization-mode are combined.
   Optimization mode tracks one int256 return value, so any boolean
   secondary invariants in the same contract are filtered out before
   the campaign — previously silent. Now emits a once-per-suite warning
   naming the optimization invariant and every dropped boolean so users
   can move them to a separate contract.

E2E tests: extend assert_all to assert the new 4/5 roll-up;
assert_all_only_primary covers the 1/2 case; new
assert_all_optimization_mode_warning verifies the warning fires with
the dropped invariant names.

Co-authored-by: Amp <amp@ampcode.com>
Amp-Thread-ID: https://ampcode.com/threads/T-019dcdd3-53f5-76b6-ac36-d59f06b58280

grandizzy committed:
feat(invariant): warn when assert_all skips invariants with persisted failures

Symmetric with the primary's existing persisted-replay warning. Echidna
and Medusa never silently drop properties between runs — properties are
re-evaluated every campaign and a previous failure doesn't suppress
them. Foundry's per-property failure file model meant secondaries with
a stale persisted counterexample were filtered out of the campaign with
no acknowledgment, so users coming from Echidna/Medusa would see fewer
invariants in the report than their contract defines.

Now emits one stderr line listing every skipped name and the cache dir
to clean, e.g.:

  Warning: test/X.t.sol:Suite: 3 invariant(s) skipped due to persisted
  failures: invariant_a, invariant_b, invariant_c. Run `forge clean`
  or delete files in cache/invariant/failures/Suite to re-include.

E2E: extends assert_all re-run case with stderr_eq snapshot asserting
the warning fires with all 3 skipped names.

Co-authored-by: Amp <amp@ampcode.com>
Amp-Thread-ID: https://ampcode.com/threads/T-019dcdd3-53f5-76b6-ac36-d59f06b58280

grandizzy committed:
Merge branch 'master' into gdzzy/issue-9727

grandizzy committed:
fix(invariant): gate afterInvariant per-run under assert_all

Previously afterInvariant was gated on failures.errors.is_empty()
campaign-wide. Under assert_all that gate stayed closed for the rest
of the campaign once any invariant broke, silently skipping the
afterInvariant hook on every subsequent run. Any assertions or
cleanup logic in afterInvariant therefore stopped running after the
first unrelated invariant failure.

Now snapshot failures.errors.len() at the start of each run and only
skip afterInvariant when the current run produced a new failure.
Preserves the legacy 'don't run afterInvariant on a run that already
failed' semantics while letting it run on subsequent runs once an
earlier invariant has broken.

E2E: new assert_all_after_invariant_runs_after_earlier_failure case
breaks invariant_first in run 1, keeps the campaign alive with a
second never-breaking invariant, and asserts an always-reverting
afterInvariant surfaces its marker in failure output.

Co-authored-by: Amp <amp@ampcode.com>
Amp-Thread-ID: https://ampcode.com/threads/T-019dce2d-57c7-734a-bbc6-6fa5e34b25de

grandizzy committed:
fix(invariant): re-evaluate secondary persisted failures on settings change

The secondary persisted-failure skip used a bare `.exists()` check at
two sites in runner.rs (the warning + InvariantContract::new filter,
and the post-campaign shrink loop). Under the new `assert_all = true`
default this meant any leftover failure file from a previous run was
treated as still valid even after the user changed a tracked setting
(target contracts/selectors, target/excluded senders, fail_on_revert),
silently dropping the secondary from the campaign with a misleading
'skipped due to persisted failures' warning.

Now both sites use the same settings-aware compatibility check the
primary's replay path uses (persisted_call_sequence settings.diff).
Stale caches fall back to a fresh evaluation; only secondaries whose
persisted settings still match the current run are honored.

Also hoists current_settings up so the new
secondary_has_compatible_persisted closure can reuse it across all
three call sites (warning, filter, shrink-loop skip).

E2E: new assert_all_secondary_persisted_revalidates_on_settings_change
runs once with fail_on_revert=false, flips it to true, re-runs and
asserts the suite roll-up shows 2/2 invariants broken — proving the
secondary was re-evaluated rather than silently filtered out.

Co-authored-by: Amp <amp@ampcode.com>
Amp-Thread-ID: https://ampcode.com/threads/T-019dce2d-57c7-734a-bbc6-6fa5e34b25de

grandizzy committed:
fix(invariant): drop hollow [FAIL] when only secondaries break under assert_all

When the selected invariant test passes but a secondary breaks under
assert_all, the report previously rendered a hollow '[FAIL]' header
(no reason, no counterexample) for the primary and the suite roll-up
overcounted broken invariants as '1 + other_failures.len()',
attributing a non-existent primary failure.

Now key the primary header on whether the primary actually broke
(`reason.is_some() || counterexample.is_some()`) and skip the
header when it didn't. Roll-up uses the same flag so the count
reflects only invariants that actually broke (e.g., 1/2 instead of
2/2). JSON shape is unchanged: top-level reason/counterexample stay
null when the selected primary didn't break, with full secondary
detail in other_failures.

E2E: new assert_all_secondary_only_failure_no_hollow_fail asserts a
secondary-only break renders 'Suite assert_all: 1/2 invariants broken'
followed by the secondary's '[FAIL: ...] <name>' block, with no
hollow primary header.

Co-authored-by: Amp <amp@ampcode.com>
Amp-Thread-ID: https://ampcode.com/threads/T-019dce2d-57c7-734a-bbc6-6fa5e34b25de

grandizzy committed:
fix(invariant): scope assert_all hollow [FAIL] suppression to secondary-only case

Previous commit suppressed the '[FAIL]' header whenever the primary had
no top-level reason or counterexample, which also matches DS-style
failures (they signal via the 'failed' flag and log events rather than
through TestResult.reason). That regressed failure_assertions::ds_style_test_failing
and test_cmd::core::legacy_assertions in CI.

Now the suppression is scoped strictly to the assert_all secondary-only
case: skip the primary header only when no primary failure AND
assert_all is in play AND there is at least one secondary to render.
DS-style, plain unit and single-invariant failures keep the original
'[FAIL]'/'[FAIL: ...]' rendering.

Co-authored-by: Amp <amp@ampcode.com>
Amp-Thread-ID: https://ampcode.com/threads/T-019dce2d-57c7-734a-bbc6-6fa5e34b25de

grandizzy committed:
fix(invariant): attribute failure event to first broken invariant in declaration order

The structured JSON 'failure' event emitted to stderr at campaign end
(consumed by benchmark and CI tooling) used 'errors.values().next()'
on a HashMap to pick its 'reason' field, while hardcoding the
'invariant' field to the primary's name. With HashMap RandomState,
the same broken set of invariants produced a different reason string
across runs, and the event was self-inconsistent (e.g.,
'invariant: invariant_balance, reason: fee miscalculation'). Three
sites used this pattern: in-run break path, afterInvariant break
path, and the preflight check fallback.

Now they walk 'invariant_contract.invariant_fns' in declaration
order (a Vec, deterministic) and pick the first one with a recorded
failure. Both 'invariant' and 'reason' fields refer to the same
function, and the event is stable across runs.

A new 'first_broken_event' helper centralizes the lookup.

E2E: assert_all_failure_event_uses_declaration_order declares three
invariants (a, b, c) that all break on the same call, runs with
'--mt invariant_c' (primary is the last declared) and asserts the
emitted event names invariant_a with reason 'a broken'.

Co-authored-by: Amp <amp@ampcode.com>
Amp-Thread-ID: https://ampcode.com/threads/T-019dce2d-57c7-734a-bbc6-6fa5e34b25de

grandizzy committed:
refactor(invariant): rename to InvariantSecondaryFailure / invariant_secondary_failures and serialize sparsely

Renames TestResult.other_failures -> invariant_secondary_failures and
the underlying InvariantOtherFailure struct -> InvariantSecondaryFailure.
The previous names were generic ('other relative to what?'); the new
names align with the existing 'primary/secondary' terminology used
throughout the assert_all rollout and follow the Rust Vec<Foo>/foos
plural-of-singular convention.

Also marks the field with #[serde(default, skip_serializing_if =
'Vec::is_empty')] so it is omitted from JSON output for any test that
has no secondary failure data — plain unit tests, fuzz tests, passing
tests. Pre-PR JSON consumers continue to see the same shape on those
results. invariant_failure_dir and assert_all_invariant_count already
had Option::is_none guards.

Updates the SimpleContractTest{NonVerbose,Verbose}.json fixtures to
drop the now-skipped empty field.

Co-authored-by: Amp <amp@ampcode.com>
Amp-Thread-ID: https://ampcode.com/threads/T-019dce2d-57c7-734a-bbc6-6fa5e34b25de

grandizzy committed:
Merge branch 'master' into gdzzy/issue-9727

0xalpharush commented on crates/evm/evm/src/executors/invariant/result.rs:
```diff
@@ -130,37 +138,50 @@ pub(crate) fn assert_invariants<FEN: FoundryEvmNetwork>(
-                calldata,
-                call_result,
-                &inner_sequence,
+        if invariant_failures.has_failure(invariant) {
```
Given this is mut (`invariant_failures: &mut InvariantFailures,`), couldn't we remove failing invariant_fn and avoid this lookup? Only other reason to keep it around would be to store duplicate failures, but we're not doing that here

grandizzy replied:
yep, good point. I evaluated the change and it's a bigger refactor, OK if will do in a follow-up / cleanup?

0xalpharush replied:
Tbh I think it is fine to just keep calling these, and we should figure out a deduplication strategy. https://github.com/foundry-rs/foundry/pull/14482#discussion_r3154539830

grandizzy replied:
kk, will follow up on #14482 with dedup strategy

grandizzy committed:
test(invariant): assert_all + fail_on_revert=false attributes assert() to all live invariants

Amp-Thread-ID: https://ampcode.com/threads/T-019dd262-ed81-723c-aaaa-8e1314bed45a
Co-authored-by: Amp <amp@ampcode.com>

grandizzy committed:
Merge branch 'master' into gdzzy/issue-9727

grandizzy committed:
Merge branch 'master' into gdzzy/issue-9727

grandizzy committed:
Merge branch 'master' into gdzzy/issue-9727

grandizzy committed:
Merge branch 'master' into gdzzy/issue-9727

grandizzy committed:
Merge branch 'master' into gdzzy/issue-9727

grandizzy committed:
Merge branch 'master' into gdzzy/issue-9727

Commit:
fix(clippy): use values() instead of iter() with unused key

Amp-Thread-ID: https://ampcode.com/threads/T-019df149-03af-722b-abda-d8ab6332ee3b
Co-authored-by: Amp <amp@ampcode.com>

Commit:
fix(clippy): drop redundant references in trimmed_hex format args

Amp-Thread-ID: https://ampcode.com/threads/T-019df149-03af-722b-abda-d8ab6332ee3b
Co-authored-by: Amp <amp@ampcode.com>

grandizzy committed:
Merge branch 'master' into gdzzy/issue-9727

0xalpharush commented on crates/evm/evm/src/executors/invariant/mod.rs:
```diff
@@ -179,6 +180,23 @@ impl InvariantFailureMetrics {
 
+/// Returns the (name, reason) of the first broken invariant in declaration order. Falls back
+/// to the primary's name with an empty reason if `failures.errors` is somehow empty.
+fn first_broken_event<'a>(
```
I think it is fine to collect only one and avoid this search

grandizzy replied:
removed in https://github.com/foundry-rs/foundry/pull/12587/changes/4bb552fd95d504f744e1944eb46d4d362e3164b4

0xalpharush commented on crates/evm/fuzz/src/invariant/mod.rs:
```diff
@@ -267,7 +267,9 @@ pub struct InvariantContract<'a> {
-    pub invariant_function: &'a Function,
+    pub primary_invariant_fn: &'a Function,
+    /// All invariant functions present in the test contract and their fail on revert config.
+    pub invariant_fns: Vec<(&'a Function, bool)>,
```
IMO this could just be one list

mablr replied:
+1

grandizzy replied:
changed in https://github.com/foundry-rs/foundry/pull/12587/changes/4bb552fd95d504f744e1944eb46d4d362e3164b4

0xalpharush commented on crates/forge/src/result.rs:
```diff
@@ -421,6 +438,26 @@ pub struct TestResult {
+    /// Each entry carries the invariant's name, the failure reason, an optional counterexample,
+    /// and the path where the counterexample has been persisted for shrinking on a subsequent run.
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub invariant_secondary_failures: Vec<InvariantSecondaryFailure>,
```
Again, I don't think we need primary vs secondary to have different data structures

grandizzy replied:
unified in https://github.com/foundry-rs/foundry/pull/12587/changes/4bb552fd95d504f744e1944eb46d4d362e3164b4 introduced concept of anchor for the current test performed

mablr commented on crates/evm/evm/src/executors/invariant/mod.rs:
```diff
@@ -695,8 +698,12 @@ impl<'a, FEN: FoundryEvmNetwork> InvariantExecutor<'a, FEN> {
 
-            // Call `afterInvariant` only if it is declared and test didn't fail already.
-            if invariant_contract.call_after_invariant && !invariant_test.has_errors() {
+            // Call `afterInvariant` only if declared and the current run produced no new
+            // failure. Under `assert_all` the campaign keeps running after earlier failures,
+            // but the hook must still execute on subsequent runs.
+            if invariant_contract.call_after_invariant
+                && invariant_test.test_data.failures.errors.len() == failures_before_run
+            {
```
Is there a risk of inconsistent `afterInvariant` hook execution after a failure for subsequent runs ?

What if an invariant fails, is the `afterInvariant` hook executed for others ?

grandizzy replied:
yep, how this works is afterInvariant is skipped only on the specific run that fails, but will still be asserted for the rest of campaign (further runs)

figtracer commented on crates/forge/src/runner.rs:
```diff
@@ -345,7 +345,9 @@ impl<'a, FEN: FoundryEvmNetwork> ContractRunner<'a, FEN> {
         // Invariant testing requires tracing to figure out what contracts were created.
         // We also want to disable `debug` for setup since we won't be using those traces.
-        let has_invariants = self.contract.abi.functions().any(|func| func.is_invariant_test());
+        let invariant_fns: Vec<_> =
```
should we validate all collected invariants before building the `InvariantContract`?  

e.g.: `invariant_bar(uint256 z)` can leak as a secondary and this fails unexpectedly

grandizzy replied:
good catch, addressed in https://github.com/foundry-rs/foundry/pull/12587/changes/0d8cc990864673b71b3e239669d051f1984d6bdf

stevencartavia commented on crates/evm/evm/src/executors/invariant/error.rs:
```diff
@@ -17,16 +18,41 @@ pub struct InvariantFailures {
     /// The latest revert reason of a run.
     pub revert_reason: Option<String>,
     /// Maps a broken invariant to its specific error.
```
```suggestion
    /// Maps each broken invariant (by function name) to its specific error.
```

grandizzy committed:
Merge branch 'master' into gdzzy/issue-9727

grandizzy committed:
Update crates/evm/evm/src/executors/invariant/error.rs

Co-authored-by: stevencartavia <112043913+stevencartavia@users.noreply.github.com>

grandizzy committed:
invariant: unify primary/secondary as one list and one failure type

The "anchor" of an invariant campaign is the test entry point selected
by `--mt` (or the per-`Test` invariant function the runner is currently
exercising). Under `assert_all`, the anchor is the campaign's named
test, and any other broken `invariant_*` functions are secondaries
shrunk and reported alongside it.

- InvariantContract: single invariant_fns list in declaration order
  with an anchor_idx pointing at the campaign's anchor invariant
- TestResult: replace primary/secondary split with invariant_failures:
  Vec<InvariantFailure>; each entry carries is_anchor so the renderer
  can omit the function name suffix when a single failure is the anchor
  (its name is already on the trailing summary line)
- assert_invariants/can_continue/assert_after_invariant return the
  broken invariant directly, removing the first_broken_event re-scan
- collapse FailedInvariantCaseData::new + with_assertion_failure into
  InvariantRunCtx::failed_case
- show invariant name on [FAIL: ...] line when not anchor or multiple
  failures (so secondaries in single-failure assert_all runs are
  still visible)

grandizzy committed:
invariant: validate signatures upstream and reject parameterized invariants

grandizzy committed:
Merge branch 'master' into gdzzy/issue-9727

mablr commented on crates/evm/evm/src/executors/invariant/mod.rs:
```diff
@@ -602,40 +588,51 @@ impl<'a, FEN: FoundryEvmNetwork> InvariantExecutor<'a, FEN> {
-                                InvariantFuzzError::Revert(case_data)
-                            });
-                            result::RichInvariantResults::new(false, None)
+                            invariant_test.set_error(
+                                anchor,
+                                if assertion_failure {
+                                    InvariantFuzzError::BrokenInvariant(case_data)
+                                } else {
+                                    InvariantFuzzError::Revert(case_data)
+                                },
+                            );
```
Only the anchor is marked as broken. Is this fine ?

grandizzy replied:
yep, that's fine, failures for the other invariants are shown inline, e.g.
- anchor invariant_a passes, invariant_b and invariant_c fail
```
Ran 1 test for test/Demo.t.sol:Demo
[FAIL: condition b broken] invariant_b
	[Sequence] (original: [..], shrunk: [..])
		sender=0x[..] addr=[src/Counter.sol:Counter]0x… calldata=inc() args=[]
		sender=0x[..] addr=[src/Counter.sol:Counter]0x… calldata=inc2() args=[]

[FAIL: condition c broken] invariant_c
	[Sequence] (original: [..], shrunk: [..])
		sender=0x[..] addr=[src/Counter.sol:Counter]0x… calldata=spike() args=[]

Suite assert_all: 2/3 invariants broken
2 invariant failure(s) persisted to cache/invariant/failures/Demo — rerun to shrink
 invariant_a() (runs: 50, calls: 4000, reverts: 12)
```
- anchor invariant_a fails AND invariant_b fails
```
Ran 1 test for test/Demo.t.sol:Demo
[FAIL: condition a broken] invariant_a
	[Sequence] (original: [..], shrunk: [..])
		sender=0x[..] addr=[src/Counter.sol:Counter]0x… calldata=inc2() args=[]
		sender=0x[..] addr=[src/Counter.sol:Counter]0x… calldata=spike() args=[]

[FAIL: condition b broken] invariant_b
	[Sequence] (original: [..], shrunk: [..])
		sender=0x[..] addr=[src/Counter.sol:Counter]0x… calldata=inc() args=[]

Suite assert_all: 2/3 invariants broken
2 invariant failure(s) persisted to cache/invariant/failures/Demo — rerun to shrink
 invariant_a() (runs: 5, calls: 273, reverts: 8)
```

mablr commented on crates/evm/evm/src/executors/invariant/mod.rs:
```diff
@@ -602,40 +588,51 @@ impl<'a, FEN: FoundryEvmNetwork> InvariantExecutor<'a, FEN> {
                         }
                         if assertion_failure || (call_result.reverted && self.config.fail_on_revert)
                         {
-                            let case_data = error::FailedInvariantCaseData::new(
-                                &invariant_contract,
-                                &self.config,
-                                &invariant_test.targeted_contracts,
-                                &current_run.inputs,
+                            // Handler-side reverts/assertion failures aren't tied to a specific
+                            // invariant body, so attribute to the campaign anchor.
+                            let anchor = invariant_contract.anchor();
+                            let case_data = error::InvariantRunCtx {
+                                contract: &invariant_contract,
+                                config: &self.config,
+                                targeted_contracts: &invariant_test.targeted_contracts,
+                                calldata: &current_run.inputs,
+                            }
+                            .failed_case(
+                                anchor,
+                                self.config.fail_on_revert,
+                                assertion_failure,
                                 call_result,
                                 &[],
-                            )
-                            .with_assertion_failure(assertion_failure);
+                            );
```
When `check_interval > 1`, the call loop skips invariant assertion on non-interval calls. For those skipped calls, this fast path seems to handle reverts and assertion failures independently from `can_continue`, diverging from the full-check path in three compounding ways under `assert_all=true`.

grandizzy replied:
good point, handler-side assertion failures are deferred to #14482 and should already handle such

mablr replied:
oh okay I see ^^

mablr reviewed (commented):
lgtm ! 🥳

0xalpharush commented:
```
Suite assert_all: 2/3 invariants broken
2 invariant failure(s) persisted to cache/invariant/failures/Demo — rerun to shrink
 invariant_a() (runs: 5, calls: 273, reverts: 8)
 ```
I'm still not sure the UX is what I'd expect but we can change this later. I think a lot of it stems from overloading `test` e.g. --mt doesn't really make sense for invariant fuzzing. I would expect to see the contract's name as the suite and each method that we've determined is an assertion test + `invariant_` as "tests". I don't really get why `invariant_a` is the "anchor" and called out here to be rerun when there are multiple failures.

grandizzy commented:
> ```
> Suite assert_all: 2/3 invariants broken
> 2 invariant failure(s) persisted to cache/invariant/failures/Demo — rerun to shrink
>  invariant_a() (runs: 5, calls: 273, reverts: 8)
> ```
> 
> I'm still not sure the UX is what I'd expect but we can change this later. I think a lot of it stems from overloading `test` e.g. --mt doesn't really make sense for invariant fuzzing. I would expect to see the contract's name as the suite and each method that we've determined is an assertion test + `invariant_` as "tests". I don't really get why `invariant_a` is the "anchor" and called out here to be rerun when there are multiple failures.

Indeed UX is not polished yet but was under the impression we want first to get this one + https://github.com/foundry-rs/foundry/issues/14448 + https://github.com/foundry-rs/foundry/pull/14565 , then get SCFuzz baseline and then iterate on top of it (both fuzzer and UX improvements). Happy to redo this and introduce new UX here, but as long as it doesn't make PR diff huge (already big IMO)

grandizzy committed:
Merge branch 'master' into gdzzy/issue-9727

forkforkdog commented:
No way you guys did it!

grandizzy commented:
> No way you guys did it!

it still needs https://github.com/foundry-rs/foundry/issues/14448 to be fully done, should land soon
